### DISCUSSION TITLE
render long text auto wrap

### DISCUSSION THREAD

**Discussion by gijdoul** [ImageMagick/ImageMagick#5837] [Help]
I have a csv file with 100 lines, each line with two texts I have to insert each line of text on an image with a specific size, font and positioning.
I succeeded with the code but I encounter a problem when the text is too long it does not go to the line and the result is bad.
I just need to auto wrap with multi line but don't know how to do with annotate my code look like this:

`convert background.jpg -gravity North -density 120 -pointsize 50 -font "Malibu"   -annotate +0+2850 "my second text"  -gravity North -pointsize 50 -font "Alegreya" -annotate +0+2650 "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged." result.jpg`

<img width="824" alt="Capture d’écran 2022-12-01 à 8 28 42 PM" src="https://user-images.githubusercontent.com/57597191/205106975-15828d90-9a2e-420d-9aab-e5320886dfcc.png">

**Comment by snibgo** [comment]
You have:
```
...  -annotate +0+2650 caption:"Lorem Ipsum ..."
```
This is wrong. It won't do what you want. You are confusing two different methods.

Use either:
```
-annotate +0+2650 "Lorem Ipsum ..."
```
... which will write text on the image that already exists in memory, or use this:
```
caption:"Lorem Ipsum ..."
```
... which will add a new image, and you can then "`-composite`" to combine your images.

**Reply by gijdoul** [reply]
yes this was last test but originaly my code was like this without caption:

`convert background.jpg -gravity North -density 120 -pointsize 50 -font "Malibu"   -annotate +0+2850 "my second text"  -gravity North -pointsize 50 -font "Alegreya" -annotate +0+2650  "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged." result.jpg`

but result remain the same

**Comment by snibgo** [comment]
What version of IM, on what platform?

For word-wrapping, I suggest using "caption:". For example:
```
magick toes.png -size 200x200 -background None caption:"Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged." -composite x2.jpg
```

![x2](https://user-images.githubusercontent.com/33812368/205138624-2ae2bb58-837a-48af-bafb-88ebdebbf22d.jpg)

**Reply by gijdoul** [reply]
I need to position the text at specific area of the image not sure I can do using "caption" 
here my two position of text using anotate but how can I do same with caption?
-annotate +0+2850
-annotate +0+2650

**Reply by fmw42** [reply]
caption: creates its own image.  So you do that to create a text image on a transparent background. Then composite it where you want over your background image. You can do that in once command if you do the caption: in a parenthesis process.

See

https://imagemagick.org/Usage/text/#caption
https://imagemagick.org/Usage/basics/#parenthesis
https://imagemagick.org/Usage/layers/#convert

**Reply by gijdoul** [reply]
I will make a try but can I choose specific font and size etc.. like I do with my original code?

**Reply by gijdoul** [reply]
> What version of IM, on what platform?
> 
Version: ImageMagick 7.1.0-51 Q16-HDRI aarch64 20499

on Imac M1

**Comment by snibgo** [comment]
> ...but how can I do same with caption?

You can use "-geometry +X+Y" before "-composite" where X and Y are the offsets (integer numbers). For example:

```
magick toes.png -size 200x200 -background None caption:"Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged." -geometry +50+70 -composite x3.jpg
```
![x3](https://user-images.githubusercontent.com/33812368/205159342-dc318f8b-f929-4e67-807b-ead457a967e3.jpg)

**Comment by urban-warrior** [comment]
For precise positioning and frequent changes in font, font color, etc., on the same page, you might consider using a Pango Markup script instead.

**Reply by gijdoul** [reply]
I never use Pango Markup, it seems "annotate" do not handle auto wrap text it's a pitty I just need this and all is working in fact I need this only when text it's long I try to add -size parameter but it's not working

**Reply by urban-warrior** [reply]
The `caption:` coder does auto wrapping of text.  You can set the caption size (width/height), set the background to transparent, auto wrap the text, and then composite the results on your original image canvas.  The caption algorithm selects the pointsize of the font to ensure the text fits within the specified width/height.

**Reply by gijdoul** [reply]
Thanks, does it work with 2 text at same time with different font and position like in my original code?
